File: server/src/protocol.rs
//! Wire protocol types. Every message that crosses the WebSocket boundary lives here.
//!
//! The shape of these types is the public contract with bot authors — keep them in sync
//! with `docs/PROTOCOL.md`. Internal simulation types should not leak into this module.

use serde::{Deserialize, Serialize};

use crate::sim::PowerupId;

/// 2D position / velocity, serialized as a 2-element JSON array `[x, y]`.
pub type Pos = [f32; 2];
pub type Vel = [f32; 2];

// ---------------------------------------------------------------------------
// Bot -> Server
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum BotMsg {
    Hello {
        name: String,
        version: String,
    },
    Ready,
    Command {
        tick: u64,
        throttle: f32,
        rudder: f32,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        fire: Option<FireCommand>,
        sensor_mode: SensorMode,
        /// One-off powerup activation for this tick. Server validates that the bot picked
        /// the powerup before the match and has not yet used it; an invalid id earns a
        /// typed `error` frame.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        activate_powerup: Option<PowerupId>,
    },
    /// Declare the (up to 2 distinct) powerups the bot will use for this match. May only
    /// be sent while the room is in `lobby` and before `ready`. Sending it twice replaces
    /// the previous selection; an invalid loadout earns a typed `error` frame and leaves
    /// the previous selection (if any) intact.
    SelectPowerups {
        powerups: Vec<PowerupId>,
    },
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
pub struct FireCommand {
    pub bearing_deg: f32,
    pub range: f32,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum SensorMode {
    Active,
    Passive,
}

// ---------------------------------------------------------------------------
// Server -> Bot
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMsg {
    Welcome {
        bot_id: String,
        ship_id: String,
        map: MapInfo,
        tick_hz: u32,
        ship_specs: ShipSpecs,
        /// Catalog of powerup ids this server understands. Bots use this to discover what
        /// they may pass to `select_powerups`. Forward-compatible: future servers may add
        /// entries; SDKs that don't recognise an id should leave it untouched.
        available_powerups: Vec<PowerupId>,
    },
    GameStart {
        tick: u64,
        starting_position: Pos,
        starting_heading_deg: f32,
    },
    Tick {
        tick: u64,
        deadline_ms: u64,
        #[serde(rename = "self")]
        self_state: SelfState,
        contacts: Vec<Contact>,
        events: Vec<TickEvent>,
    },
    GameOver {
        winner: Option<String>,
        final_tick: u64,
        replay_id: String,
    },
    /// Emitted by the room when it transitions back to the lobby after a match. Bots
    /// that wish to participate in the next match should re-send `ready`. `tick` is
    /// always `0` and is included only to give the message a numeric field for parity
    /// with `GameStart`.
    Lobby {
        tick: u64,
    },
    Error {
        code: String,
        message: String,
    },
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
pub struct MapInfo {
    pub width: u32,
    pub height: u32,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
pub struct ShipSpecs {
    pub max_forward_speed: f32,
    pub max_reverse_speed: f32,
    pub acceleration: f32,
    pub turn_rate_deg_per_s: f32,
    pub hull_hp: u32,
    pub max_ammo: u32,
    pub gun_cooldown_ticks: u32,
    pub hit_radius: f32,
    pub shell_speed: f32,
    pub max_shell_range: f32,
    pub splash_radius: f32,
    pub max_splash_damage: u32,
}

impl ShipSpecs {
    /// Derive the `welcome`-payload spec from a match's [`SimConfig`]. Bots base their
    /// physics models on what we send here, so this must always reflect the parameters the
    /// simulation will actually run with.
    pub fn from_config(config: &crate::sim::SimConfig) -> ShipSpecs {
        ShipSpecs {
            max_forward_speed: config.max_forward_speed,
            max_reverse_speed: config.max_reverse_speed,
            acceleration: config.acceleration,
            turn_rate_deg_per_s: config.turn_rate_deg_per_s,
            hull_hp: config.hull_hp,
            max_ammo: config.max_ammo,
            gun_cooldown_ticks: config.gun_cooldown_ticks,
            hit_radius: config.hit_radius,
            shell_speed: config.shell_speed,
            max_shell_range: config.max_shell_range,
            splash_radius: config.splash_radius,
            max_splash_damage: config.max_splash_damage,
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct SelfState {
    pub pos: Pos,
    pub heading_deg: f32,
    pub speed: f32,
    pub hp: u32,
    pub ammo: u32,
    pub rudder: f32,
    pub throttle: f32,
    /// Loadout the bot picked for the match (in pick order, up to 2 entries). Sent on
    /// every tick so a bot reconnecting after a brief drop can rehydrate without state.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub selected_powerups: Vec<PowerupId>,
    /// One entry per picked powerup, in the same order as `selected_powerups`.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub powerup_status: Vec<PowerupStatus>,
}

/// Live status for one of the bot's picked powerups. Sent inside `tick.self.powerup_status`.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub struct PowerupStatus {
    pub id: PowerupId,
    /// True once the bot has activated this powerup; activating a used powerup is rejected.
    pub used: bool,
    /// Ticks remaining of the active effect. `0` when not currently active.
    pub active_ticks_left: u32,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct Contact {
    pub id: String,
    pub kind: ContactKind,
    pub pos: Pos,
    pub bearing_deg: f32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub range: Option<f32>,
    pub confidence: f32,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ContactKind {
    Ship,
    Shell,
    Unknown,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum TickEvent {
    Hit {
        amount: u32,
    },
    ShellSplash {
        pos: Pos,
    },
    /// A powerup was activated. Sent to the activating bot for every own activation
    /// (`contact_id` is `null` — you are not a contact to yourself), and to other bots
    /// only when the activating ship actually appeared in that viewer's sensor sweep this
    /// tick — tagged with the same per-tick anonymized `c_<n>` contact id the bot sees in
    /// `contacts`. Never carries a ground-truth `ship_id`: like contacts, the event is
    /// re-anonymized every tick so a bot can't track a specific opponent across ticks.
    /// Counter-battery trace reveals come in via regular `contacts`, not this event.
    PowerupActivated {
        #[serde(skip_serializing_if = "Option::is_none")]
        contact_id: Option<String>,
        powerup: PowerupId,
    },
}

// ---------------------------------------------------------------------------
// Server -> Spectator
// ---------------------------------------------------------------------------

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SpectatorMsg {
    World {
        tick: u64,
        ships: Vec<SpectatorShip>,
        shells: Vec<SpectatorShell>,
        events: Vec<SpectatorEvent>,
        /// Live smoke clouds on the battlefield. Empty when no `smoke_screen` is active.
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        smoke_clouds: Vec<SpectatorSmokeCloud>,
        /// Live decoy phantoms.
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        decoys: Vec<SpectatorDecoy>,
    },
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct SpectatorShip {
    pub id: String,
    pub bot_name: String,
    pub pos: Pos,
    pub heading_deg: f32,
    /// Signed scalar speed: positive = ahead, negative = reverse.
    pub speed: f32,
    pub hp: u32,
    pub ammo: u32,
    /// Last commanded throttle in `[-1, 1]`.
    pub throttle: f32,
    /// Last commanded rudder in `[-1, 1]`.
    pub rudder: f32,
    pub alive: bool,
    /// Lobby readiness flag. `true` once the bot has sent `ready`.
    pub ready: bool,
    /// Commands accepted by the room over the most recent rolling 1-second window
    /// (1s of *sim* time, i.e. `tick_hz` ticks). Stays 0 while the room is in lobby.
    pub commands_per_sec: f32,
    /// Last commanded sensor mode. Used by the renderer to draw an active-radar ring.
    pub sensor_mode: SensorMode,
    /// Powerups this bot picked for the match (in pick order).
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub selected_powerups: Vec<PowerupId>,
    /// Per-pick status for the spectator HUD: which are used, which are active, and how
    /// many ticks of effect remain.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub powerup_status: Vec<PowerupStatus>,
}

/// One smoke cloud on the spectator wire.
#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
pub struct SpectatorSmokeCloud {
    pub pos: Pos,
    pub radius: f32,
    pub expires_at: u64,
}

/// One decoy phantom on the spectator wire.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct SpectatorDecoy {
    pub fake_id: u32,
    pub owner: String,
    pub pos: Pos,
    pub heading_deg: f32,
    pub expires_at: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
pub struct SpectatorShell {
    pub id_index: u32,
    pub pos: Pos,
    pub vel: Vel,
    pub ttl_ticks: u32,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SpectatorEvent {
    Hit { ship_id: String, amount: u32 },
    ShellSplash { pos: Pos },
    Death { ship_id: String },
    PowerupActivated { ship_id: String, powerup: PowerupId },
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Standardized error codes sent inside `ServerMsg::Error`. Keep this list small and
/// document each code in `docs/PROTOCOL.md`.
pub mod error_code {
    pub const MALFORMED_JSON: &str = "malformed_json";
    pub const INVALID_MESSAGE: &str = "invalid_message";
    pub const BINARY_FRAMES_UNSUPPORTED: &str = "binary_frames_unsupported";
    pub const TOO_MANY_VIOLATIONS: &str = "too_many_violations";
    pub const LATE_COMMAND: &str = "late_command";
    pub const COOLDOWN_ACTIVE: &str = "cooldown_active";
    pub const NO_AMMO: &str = "no_ammo";
    pub const INVALID_NAME: &str = "invalid_name";
    pub const DUPLICATE_NAME: &str = "duplicate_name";
    pub const STALE_COMMAND: &str = "stale_command";
    pub const NON_FINITE_VALUE: &str = "non_finite_value";
    pub const HANDSHAKE_TIMEOUT: &str = "handshake_timeout";
    pub const CONNECTION_LIMIT: &str = "connection_limit";
    // --- Powerups ----------------------------------------------------------
    pub const POWERUP_UNKNOWN: &str = "powerup_unknown";
    pub const POWERUP_DUPLICATE: &str = "powerup_duplicate";
    pub const POWERUP_WRONG_COUNT: &str = "powerup_wrong_count";
    pub const POWERUP_LOBBY_ONLY: &str = "powerup_lobby_only";
    pub const POWERUP_NOT_SELECTED: &str = "powerup_not_selected";
    pub const POWERUP_ALREADY_USED: &str = "powerup_already_used";
}

/// Maximum length of a bot's `hello.name`. Names are also restricted to
/// `[A-Za-z0-9 _-]`; see [`validate_bot_name`].
pub const MAX_BOT_NAME_LEN: usize = 32;

/// Returns `Ok(())` if `name` is a valid bot identifier:
/// - 1..=`MAX_BOT_NAME_LEN` characters,
/// - every byte is ASCII alphanumeric, space, underscore, or hyphen.
///
/// Restricting the charset keeps replay logs greppable and stops spectator-UI injection
/// via unicode controls / RTL overrides.
pub fn validate_bot_name(name: &str) -> Result<(), &'static str> {
    if name.is_empty() {
        return Err("name must not be empty");
    }
    if name.len() > MAX_BOT_NAME_LEN {
        return Err("name exceeds 32 bytes");
    }
    if !name
        .bytes()
        .all(|b| b.is_ascii_alphanumeric() || b == b' ' || b == b'_' || b == b'-')
    {
        return Err("name may only contain A-Z, a-z, 0-9, space, underscore, or hyphen");
    }
    Ok(())
}

/// Build a `ServerMsg::Error`.
pub fn error_msg(code: &str, message: impl Into<String>) -> ServerMsg {
    ServerMsg::Error {
        code: code.to_string(),
        message: message.into(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn roundtrip<T>(value: &T)
    where
        T: Serialize + for<'de> Deserialize<'de> + PartialEq + std::fmt::Debug,
    {
        let json = serde_json::to_string(value).expect("serialize");
        let parsed: T = serde_json::from_str(&json).expect("deserialize");
        assert_eq!(value, &parsed, "roundtrip mismatch: {json}");
    }

    #[test]
    fn bot_msg_roundtrips() {
        roundtrip(&BotMsg::Hello {
            name: "captain_kirk".into(),
            version: "1.0".into(),
        });
        roundtrip(&BotMsg::Ready);
        roundtrip(&BotMsg::Command {
            tick: 142,
            throttle: 0.8,
            rudder: -0.3,
            fire: Some(FireCommand {
                bearing_deg: 47.5,
                range: 300.0,
            }),
            sensor_mode: SensorMode::Active,
            activate_powerup: None,
        });
        roundtrip(&BotMsg::Command {
            tick: 143,
            throttle: 0.0,
            rudder: 0.0,
            fire: None,
            sensor_mode: SensorMode::Passive,
            activate_powerup: Some(PowerupId::Overdrive),
        });
        roundtrip(&BotMsg::SelectPowerups {
            powerups: vec![PowerupId::RapidFire, PowerupId::HeavyShell],
        });
    }

    #[test]
    fn server_msg_roundtrips() {
        roundtrip(&ServerMsg::Welcome {
            bot_id: "b_3".into(),
            ship_id: "s_3".into(),
            map: MapInfo {
                width: 1000,
                height: 1000,
            },
            tick_hz: 10,
            ship_specs: ShipSpecs::from_config(&crate::sim::SimConfig::default()),
            available_powerups: PowerupId::all().to_vec(),
        });
        roundtrip(&ServerMsg::GameStart {
            tick: 0,
            starting_position: [120.0, 340.0],
            starting_heading_deg: 90.0,
        });
        roundtrip(&ServerMsg::Tick {
            tick: 142,
            deadline_ms: 80,
            self_state: SelfState {
                pos: [203.4, 511.7],
                heading_deg: 92.3,
                speed: 4.1,
                hp: 78,
                ammo: 14,
                rudder: -0.3,
                throttle: 0.8,
                selected_powerups: vec![PowerupId::Overdrive, PowerupId::RapidFire],
                powerup_status: vec![
                    PowerupStatus {
                        id: PowerupId::Overdrive,
                        used: false,
                        active_ticks_left: 0,
                    },
                    PowerupStatus {
                        id: PowerupId::RapidFire,
                        used: true,
                        active_ticks_left: 12,
                    },
                ],
            },
            contacts: vec![Contact {
                id: "c_a1".into(),
                kind: ContactKind::Ship,
                pos: [450.0, 510.0],
                bearing_deg: 88.0,
                range: Some(247.0),
                confidence: 0.85,
            }],
            events: vec![
                TickEvent::Hit { amount: 12 },
                TickEvent::ShellSplash {
                    pos: [220.0, 505.0],
                },
            ],
        });
        roundtrip(&ServerMsg::GameOver {
            winner: Some("b_3".into()),
            final_tick: 1843,
            replay_id: "match_20260508_171203".into(),
        });
        roundtrip(&ServerMsg::GameOver {
            winner: None,
            final_tick: 3000,
            replay_id: "match_draw".into(),
        });
        roundtrip(&error_msg(
            error_code::LATE_COMMAND,
            "command for tick 142 arrived after deadline",
        ));
    }

    #[test]
    fn spectator_msg_roundtrips() {
        roundtrip(&SpectatorMsg::World {
            tick: 142,
            ships: vec![SpectatorShip {
                id: "s_1".into(),
                bot_name: "captain_kirk".into(),
                pos: [203.4, 511.7],
                heading_deg: 92.3,
                speed: 4.1,
                hp: 78,
                ammo: 14,
                throttle: 0.8,
                rudder: -0.3,
                alive: true,
                ready: true,
                commands_per_sec: 10.0,
                sensor_mode: SensorMode::Active,
                selected_powerups: vec![PowerupId::SmokeScreen, PowerupId::AwacsScan],
                powerup_status: vec![
                    PowerupStatus {
                        id: PowerupId::SmokeScreen,
                        used: true,
                        active_ticks_left: 0,
                    },
                    PowerupStatus {
                        id: PowerupId::AwacsScan,
                        used: false,
                        active_ticks_left: 0,
                    },
                ],
            }],
            shells: vec![SpectatorShell {
                id_index: 22,
                pos: [310.0, 500.0],
                vel: [40.0, 5.0],
                ttl_ticks: 18,
            }],
            events: vec![
                SpectatorEvent::Hit {
                    ship_id: "s_1".into(),
                    amount: 12,
                },
                SpectatorEvent::ShellSplash {
                    pos: [220.0, 505.0],
                },
                SpectatorEvent::Death {
                    ship_id: "s_2".into(),
                },
                SpectatorEvent::PowerupActivated {
                    ship_id: "s_1".into(),
                    powerup: PowerupId::SmokeScreen,
                },
            ],
            smoke_clouds: vec![SpectatorSmokeCloud {
                pos: [203.4, 511.7],
                radius: 60.0,
                expires_at: 222,
            }],
            decoys: vec![],
        });
    }

    #[test]
    fn parses_canonical_command_from_spec() {
        // Lifted verbatim from system-design.md §4.1
        let json = r#"{
            "type": "command",
            "tick": 142,
            "throttle": 0.8,
            "rudder": -0.3,
            "fire": { "bearing_deg": 47.5, "range": 300.0 },
            "sensor_mode": "active"
        }"#;
        let parsed: BotMsg = serde_json::from_str(json).expect("parse");
        match parsed {
            BotMsg::Command {
                tick,
                throttle,
                rudder,
                fire,
                sensor_mode,
                activate_powerup,
            } => {
                assert_eq!(tick, 142);
                assert!((throttle - 0.8).abs() < 1e-6);
                assert!((rudder - -0.3).abs() < 1e-6);
                assert_eq!(
                    fire,
                    Some(FireCommand {
                        bearing_deg: 47.5,
                        range: 300.0
                    })
                );
                assert_eq!(sensor_mode, SensorMode::Active);
                assert!(activate_powerup.is_none());
            }
            other => panic!("expected Command, got {other:?}"),
        }
    }

    #[test]
    fn rejects_unknown_message_type() {
        let json = r#"{"type":"explode_now"}"#;
        let result: Result<BotMsg, _> = serde_json::from_str(json);
        assert!(result.is_err(), "should reject unknown message type");
    }

    #[test]
    fn rejects_empty_object() {
        let result: Result<BotMsg, _> = serde_json::from_str("{}");
        assert!(result.is_err(), "empty object has no `type` discriminant");
    }

    #[test]
    fn validate_bot_name_accepts_typical_names() {
        for ok in ["alice", "Bot_42", "captain kirk", "ship-1", "A"] {
            assert!(validate_bot_name(ok).is_ok(), "should accept {ok:?}");
        }
    }

    #[test]
    fn validate_bot_name_rejects_garbage() {
        assert!(validate_bot_name("").is_err(), "empty name rejected");
        let too_long = "x".repeat(MAX_BOT_NAME_LEN + 1);
        assert!(
            validate_bot_name(&too_long).is_err(),
            "33-byte name rejected"
        );
        for bad in ["alice\n", "bob\t", "🛳️ ship", "<script>", "../etc/passwd"] {
            assert!(validate_bot_name(bad).is_err(), "should reject {bad:?}");
        }
    }
}

File: sdk-python/naval_sdk/protocol.py
"""Typed views of the wire protocol.

These dataclasses mirror the messages documented in `docs/PROTOCOL.md`. They are
*derived* from raw dicts that arrive over the WebSocket — bot authors are free to
ignore them and read raw frames via `Bot.raw_recv()` / `Bot.raw_send()`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional, Sequence, Tuple

from .helpers import Point, bearing_to, lead_target

SensorMode = Literal["active", "passive"]
ContactKind = Literal["ship", "shell", "unknown"]


@dataclass(frozen=True)
class ShipSpecs:
    max_forward_speed: float
    max_reverse_speed: float
    acceleration: float
    turn_rate_deg_per_s: float
    hull_hp: int
    max_ammo: int
    gun_cooldown_ticks: int
    hit_radius: float
    shell_speed: float
    max_shell_range: float
    splash_radius: float
    max_splash_damage: int

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "ShipSpecs":
        return cls(
            max_forward_speed=float(d["max_forward_speed"]),
            max_reverse_speed=float(d["max_reverse_speed"]),
            acceleration=float(d["acceleration"]),
            turn_rate_deg_per_s=float(d["turn_rate_deg_per_s"]),
            hull_hp=int(d["hull_hp"]),
            max_ammo=int(d["max_ammo"]),
            gun_cooldown_ticks=int(d["gun_cooldown_ticks"]),
            hit_radius=float(d["hit_radius"]),
            shell_speed=float(d["shell_speed"]),
            max_shell_range=float(d["max_shell_range"]),
            splash_radius=float(d["splash_radius"]),
            max_splash_damage=int(d["max_splash_damage"]),
        )


@dataclass(frozen=True)
class MapInfo:
    width: int
    height: int

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "MapInfo":
        return cls(width=int(d["width"]), height=int(d["height"]))


@dataclass(frozen=True)
class Welcome:
    bot_id: str
    ship_id: str
    map: MapInfo
    tick_hz: int
    ship_specs: ShipSpecs
    #: Powerup ids the server understands. Forward-compatible: unknown future entries
    #: are simply passed back to the server if a bot picks them — the server validates.
    available_powerups: List[str] = field(default_factory=list)

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Welcome":
        return cls(
            bot_id=str(d["bot_id"]),
            ship_id=str(d["ship_id"]),
            map=MapInfo.from_dict(d["map"]),
            tick_hz=int(d["tick_hz"]),
            ship_specs=ShipSpecs.from_dict(d["ship_specs"]),
            available_powerups=[str(p) for p in d.get("available_powerups", [])],
        )


@dataclass(frozen=True)
class PowerupStatus:
    """Live status for one of the bot's picked powerups, mirrored in every tick."""

    id: str
    used: bool
    active_ticks_left: int

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "PowerupStatus":
        return cls(
            id=str(d["id"]),
            used=bool(d.get("used", False)),
            active_ticks_left=int(d.get("active_ticks_left", 0)),
        )


@dataclass(frozen=True)
class GameStart:
    tick: int
    starting_position: Tuple[float, float]
    starting_heading_deg: float

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "GameStart":
        pos = d["starting_position"]
        return cls(
            tick=int(d["tick"]),
            starting_position=(float(pos[0]), float(pos[1])),
            starting_heading_deg=float(d["starting_heading_deg"]),
        )


@dataclass(frozen=True)
class SelfState:
    pos: Tuple[float, float]
    heading_deg: float
    speed: float
    hp: int
    ammo: int
    rudder: float
    throttle: float
    #: Loadout the bot picked for the match, in pick order. Empty if the bot never sent
    #: `select_powerups`.
    selected_powerups: Tuple[str, ...] = ()
    #: One entry per picked powerup, same order. Use this to check whether a powerup is
    #: still available, currently active, or already used.
    powerup_status: Tuple[PowerupStatus, ...] = ()

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "SelfState":
        pos = d["pos"]
        return cls(
            pos=(float(pos[0]), float(pos[1])),
            heading_deg=float(d["heading_deg"]),
            speed=float(d["speed"]),
            hp=int(d["hp"]),
            ammo=int(d["ammo"]),
            rudder=float(d["rudder"]),
            throttle=float(d["throttle"]),
            selected_powerups=tuple(str(p) for p in d.get("selected_powerups", [])),
            powerup_status=tuple(
                PowerupStatus.from_dict(s) for s in d.get("powerup_status", [])
            ),
        )

    # ---- Convenience for bots ---------------------------------------------

    def powerup(self, powerup_id: str) -> Optional[PowerupStatus]:
        """Look up the live status of a specific picked powerup, or `None` if not picked."""
        for status in self.powerup_status:
            if status.id == powerup_id:
                return status
        return None

    def powerup_ready(self, powerup_id: str) -> bool:
        """True iff the bot picked this powerup and has not yet activated it."""
        status = self.powerup(powerup_id)
        return status is not None and not status.used

    def powerup_active(self, powerup_id: str) -> bool:
        """True iff this powerup is currently in effect (`active_ticks_left > 0`)."""
        status = self.powerup(powerup_id)
        return status is not None and status.active_ticks_left > 0


@dataclass(frozen=True)
class Contact:
    id: str
    kind: ContactKind
    pos: Tuple[float, float]
    bearing_deg: float
    range: Optional[float]
    confidence: float

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Contact":
        pos = d["pos"]
        rng = d.get("range")
        return cls(
            id=str(d["id"]),
            kind=d.get("kind", "unknown"),
            pos=(float(pos[0]), float(pos[1])),
            bearing_deg=float(d["bearing_deg"]),
            range=None if rng is None else float(rng),
            confidence=float(d.get("confidence", 0.0)),
        )


@dataclass(frozen=True)
class HitEvent:
    amount: int


@dataclass(frozen=True)
class ShellSplashEvent:
    pos: Tuple[float, float]


@dataclass(frozen=True)
class PowerupActivatedEvent:
    """Reported when a ship activates a powerup. Always emitted for the bot's own
    activations (``contact_id`` is ``None`` — you are not a contact to yourself); emitted
    for another ship only when that ship shows up in this bot's sensor sweep this tick, in
    which case ``contact_id`` is the same per-tick anonymized ``c_<n>`` id it appears under
    in ``contacts``. Never the ground-truth ship id — the event is re-anonymized every
    tick, so it can't be used to track a specific opponent across ticks."""

    contact_id: Optional[str]
    powerup: str


TickEvent = Any  # HitEvent | ShellSplashEvent | PowerupActivatedEvent | unknown


def _parse_event(d: Dict[str, Any]) -> TickEvent:
    try:
        kind = d.get("type")
        if kind == "hit":
            return HitEvent(amount=int(d["amount"]))
        if kind == "shell_splash":
            pos = d["pos"]
            return ShellSplashEvent(pos=(float(pos[0]), float(pos[1])))
        if kind == "powerup_activated":
            raw_contact = d.get("contact_id")
            return PowerupActivatedEvent(
                contact_id=None if raw_contact is None else str(raw_contact),
                powerup=str(d["powerup"]),
            )
    except (AttributeError, KeyError, IndexError, TypeError, ValueError):
        # A known event type arrived malformed (missing/ill-typed field). Fall
        # through to the raw-dict path so one bad event can't sink the whole tick.
        return d
    return d  # forward-compatible: unknown event types stay as raw dicts


@dataclass(frozen=True)
class WorldView:
    """Bot-side view of a single `tick` message."""

    tick: int
    deadline_ms: int
    self_state: SelfState
    contacts: List[Contact]
    events: List[TickEvent]

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "WorldView":
        return cls(
            tick=int(d["tick"]),
            deadline_ms=int(d["deadline_ms"]),
            self_state=SelfState.from_dict(d["self"]),
            contacts=[Contact.from_dict(c) for c in d.get("contacts", [])],
            events=[_parse_event(e) for e in d.get("events", [])],
        )

    # Convenience accessors -------------------------------------------------

    @property
    def me(self) -> SelfState:
        return self.self_state

    def nearest_contact(self) -> Optional[Contact]:
        ranged = [c for c in self.contacts if c.range is not None]
        if not ranged:
            return None
        return min(ranged, key=lambda c: c.range)  # type: ignore[arg-type,return-value]


@dataclass(frozen=True)
class GameOver:
    winner: Optional[str]
    final_tick: int
    replay_id: str

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "GameOver":
        return cls(
            winner=d.get("winner"),
            final_tick=int(d["final_tick"]),
            replay_id=str(d["replay_id"]),
        )


# ---------------------------------------------------------------------------
# Outbound
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class FireCommand:
    bearing_deg: float
    range: float

    def to_dict(self) -> Dict[str, Any]:
        return {"bearing_deg": float(self.bearing_deg), "range": float(self.range)}


@dataclass
class Command:
    """A bot's response to a single `tick`.

    `throttle` and `rudder` are clamped to `[-1, 1]` by the server. `sensor_mode`
    defaults to `"active"`. To fire, either pass `fire=FireCommand(...)` directly
    or call `Command.fire_at(target_pos)` / `.fire_at(target_pos, target_vel, ...)`.

    To activate one of your picked powerups, set `activate_powerup="overdrive"` (or any
    other id from `welcome.available_powerups`). The server validates that the powerup
    is in your loadout and hasn't been used yet — a bad id earns a typed `error` frame.
    """

    throttle: float = 0.0
    rudder: float = 0.0
    sensor_mode: SensorMode = "active"
    fire: Optional[FireCommand] = None
    activate_powerup: Optional[str] = None

    def fire_at(
        self,
        target_pos: Point,
        *,
        shooter_pos: Optional[Point] = None,
        target_vel: Optional[Sequence[float]] = None,
        shell_speed: float = 70.0,
        range: Optional[float] = None,
        lead: bool = True,
    ) -> "Command":
        """Aim a shell at `target_pos`.

        If `shooter_pos` is supplied and `lead=True` and `target_vel` is non-zero,
        the SDK leads the target using `shell_speed` (default 70, matching the
        server's `ship_specs.shell_speed`). Prefer passing the value you read
        from `welcome.ship_specs.shell_speed` so your lead math tracks any
        future balance changes. When `shooter_pos` is omitted, the bearing is
        computed from the origin — pass `shooter_pos=view.me.pos` to do it
        right.

        `range` defaults to the distance from `shooter_pos` to the aim point,
        clamped server-side to `max_shell_range`.
        """
        origin = tuple(shooter_pos) if shooter_pos is not None else (0.0, 0.0)
        aim: Tuple[float, float] = (float(target_pos[0]), float(target_pos[1]))
        if lead and target_vel is not None and (target_vel[0] != 0.0 or target_vel[1] != 0.0):
            predicted = lead_target(origin, aim, target_vel, shell_speed)
            if predicted is not None:
                aim = predicted

        bearing = bearing_to(origin, aim)
        if range is None:
            dx = aim[0] - origin[0]
            dy = aim[1] - origin[1]
            range_val = (dx * dx + dy * dy) ** 0.5
        else:
            range_val = float(range)

        self.fire = FireCommand(bearing_deg=bearing, range=range_val)
        return self

    def to_dict(self, tick: int) -> Dict[str, Any]:
        out: Dict[str, Any] = {
            "type": "command",
            "tick": int(tick),
            "throttle": float(self.throttle),
            "rudder": float(self.rudder),
            "sensor_mode": self.sensor_mode,
        }
        if self.fire is not None:
            out["fire"] = self.fire.to_dict()
        if self.activate_powerup is not None:
            out["activate_powerup"] = str(self.activate_powerup)
        return out
